### PR TITLE
feat: adds members list to org page

### PR DESCRIPTION
**Pull request by inpaceee**
#920

### CODE REVIEW

**Comment by inpaceee** [comment]
still can't view organization as a project member

File: packages/backend/src/graphql/user/user.ts [hunk 1]
```diff
@@ -11,21 +13,48 @@ export const User = builder.prismaObject('User', {
...
-      authScopes: (_user, { projectId }) => ({ isMemberByProject: projectId.toString() }),
-      description: 'Role of the user in a project',
-      resolve: async (user, { projectId }) => {
-        const projectMembership = await prisma.projectMembership.findUniqueOrThrow({
-          select: { role: true },
-          where: {
-            userId_projectId: {
-              userId: user.id,
-              projectId: projectId.toString(),
+      args: {
```
Comment (moncapitaine) [inline packages/backend/src/graphql/user/user.ts:16]: split it into 2 fields for the user:
1. projectRole
2. organizationRole

File: packages/backend/src/graphql/projectMembership/mutations/projectMembershipJoinMutationField.ts [hunk 2]
```diff
@@ -37,6 +44,24 @@ builder.mutationField('projectMembershipJoin', (t) =>
         update: {},
       })
 
+      // If the project belongs to an organization, add the user to the organization
```
Comment (moncapitaine) [inline packages/backend/src/graphql/projectMembership/mutations/projectMembershipJoinMutationField.ts:47]: remove the comment. This is obvious.

File: packages/backend/src/graphql/project/mutations/projectUpdateMutationField.ts [hunk 3]
```diff
@@ -37,6 +37,33 @@ builder.mutationField('projectUpdate', (t) =>
           organizationId,
         },
       })
+
+      if (organizationId) {
```
Comment (moncapitaine) [inline packages/backend/src/graphql/project/mutations/projectUpdateMutationField.ts:41]:
```suggestion
      if (!organizationId) { ....return
```

File: packages/backend/src/graphql/project/mutations/projectUpdateMutationField.ts [hunk 4]
```diff
@@ -37,6 +37,33 @@ builder.mutationField('projectUpdate', (t) =>
...
+        const projectMemberships = await prisma.projectMembership.findMany({
+          where: { projectId: id.toString() },
+        })
+
+        await Promise.all(
+          projectMemberships.map(async (membership) => {
+            await prisma.organizationMembership.upsert({
```
Comment (moncapitaine) [inline packages/backend/src/graphql/project/mutations/projectUpdateMutationField.ts:48]: can we use createMany instead? Create all users in organizationMemberships that are members in projectMemberships that do not exist in organization

File: packages/backend/src/graphql/organization/queries/organizationsQueryField.ts [hunk 5]
```diff
@@ -14,12 +14,27 @@ builder.queryField('organizations', (t) =>
...
+              ...getWhereFromOrganizationFilter(filter),
+              organizationMemberships: {
+                some: {
+                  userId: context.session.user.id,
+                },
+              },
             },
-          },
+            {
+              projects: {
```
Comment (moncapitaine) [inline packages/backend/src/graphql/organization/queries/organizationsQueryField.ts:27]: this shows orgs even if the user is not an org member.

File: packages/backend/src/graphql/organization/organization.ts [hunk 6]
```diff
@@ -17,16 +17,68 @@ export const Organization = builder.prismaObject('Organization', {
       select: { archivedAt: true },
       resolve: (organization) => !!organization.archivedAt,
     }),
+    role: t.withAuth({ isLoggedIn: true }).string({
```
Comment (moncapitaine) [inline packages/backend/src/graphql/organization/organization.ts:20]: we do not use this. Can be deleted.

File: packages/backend/src/graphql/organization/organization.ts [hunk 7]
```diff
@@ -17,16 +17,68 @@ export const Organization = builder.prismaObject('Organization', {
...
+      type: ['User'],
+      args: {
+        includeProjectMembers: t.arg.boolean({
+          defaultValue: false,
+          description:
+            'Set this to true if you want to also see the users who are members of projects in the organization',
+        }),
       },
+      resolve: (query, organization, { includeProjectMembers }) =>
+        prisma.user.findMany({
```
Comment (moncapitaine) [inline packages/backend/src/graphql/organization/organization.ts:68]: only query for org memberships in the db, no project memberships needed here any more.

File: packages/backend/src/graphql/builder.ts [hunk 8]
```diff
@@ -128,6 +144,10 @@ export const builder = new SchemaBuilder<{
             userId: context.session.user.id,
             organizationId,
           },
+          role: 'ADMIN',
+        },
+        select: {
```
Comment (moncapitaine) [inline packages/backend/src/graphql/builder.ts:149]: remove line

File: apps/web/frontend/components/organizationMemberList/addUserToOrganizationButton.tsx
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/addUserToOrganizationButton.tsx:1]: Button not needed any more.

File: apps/web/frontend/components/organizationMemberList/organizationMemberList.tsx [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+              {session.data?.user.id !== user.id &&
+                (organization.canModify ? (
+                  <RemoveUserFromOrganizationButton user={user} organization={organization} />
+                ) : (
+                  <AddUserToOrganizationButton user={user} organization={organization} />
+                ))}
+            </td>
+          </tr>
+        ))}
+        {projects.map((project) =>
```
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/organizationMemberList.tsx:119]: not needed.

File: apps/web/frontend/components/organizationMemberList/removeUserFromOrganizationButton.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      userId: user.id,
+    })
+  }
+
+  return (
+    <>
+      <button
+        className="btn btn-error btn-sm"
+        onClick={() => dialogReference.current?.showModal()}
+        title="Remove user from the organization"
```
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/removeUserFromOrganizationButton.tsx:52]: "Remove user from the organization" plus from all projects of this organization.

File: apps/web/frontend/components/organizationMemberList/roleButton.tsx
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/roleButton.tsx:1]: re-use the component from project instead.

File: apps/web/frontend/components/organizationMemberList/roleButton.tsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import { Role } from '../../mocks/mocks.generated'
```
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/roleButton.tsx:1]: graphql generated not from MSW

File: apps/web/frontend/components/organizationMemberList/roleLabel.tsx [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { Role } from '../../generated/gql/graphql'
+
+interface RoleLabelProps {
+  role: Role
+  isOrganizationMember: boolean
```
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/roleLabel.tsx:5]: not needed any more.

File: apps/web/frontend/components/organizationMemberList/roleLabel.tsx [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+export const RoleLabel = ({ role, isOrganizationMember }: RoleLabelProps) => {
+  if (role === Role.Admin && isOrganizationMember) {
+    return <span className="badge badge-primary">Organization admin</span>
+  }
+  if (isOrganizationMember) {
+    return <span className="badge badge-ghost">Organization member</span>
+  }
+  return <span className="badge badge-ghost">Project member</span>
```
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/roleLabel.tsx:15]: not needed.

File: apps/web/frontend/components/organizationMemberList/roleLabel.tsx
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/roleLabel.tsx:1]: rename organizationRoleLabel.tsx

File: apps/web/frontend/components/organizationMemberList/roleLabel.tsx [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import { Role } from '../../generated/gql/graphql'
+
+interface RoleLabelProps {
+  role: Role
+  isOrganizationMember: boolean
+}
+
+export const RoleLabel = ({ role, isOrganizationMember }: RoleLabelProps) => {
```
Comment (moncapitaine) [inline apps/web/frontend/components/organizationMemberList/roleLabel.tsx:8]:
```suggestion
export const OrganizationRoleLabel = ({ role, isOrganizationMember }: RoleLabelProps) => {
```

File: packages/backend/src/graphql/user/user.test.ts
Comment (PascalHelbig) [inline packages/backend/src/graphql/user/user.test.ts:1]: move and rename to `usersQueryField.test.ts`

File: packages/backend/src/graphql/project/project.ts [hunk 15]
```diff
@@ -61,28 +61,28 @@ export const Project = builder.prismaObject('Project', {
...
       select: { id: true },
       resolve: async (project, _arguments, context) => {
         const projectMembership = await prisma.projectMembership.findUnique({
-          select: { role: true },
+          select: { projectRole: true },
           where: { userId_projectId: { projectId: project.id, userId: context.session.user.id } },
         })
 
-        return projectMembership?.role === 'ADMIN'
+        return projectMembership?.projectRole === 'ADMIN'
```
Comment (PascalHelbig) [inline packages/backend/src/graphql/project/project.ts:85]: should also consider org admins

**Review by PascalHelbig — APPROVED** [review]

**Pull request merged**